tests/Integration/inc/classes/preload/Homepage/preload.php

```diff
@@ -0,0 +1,146 @@
+use WP_Rocket\Tests\Integration\Fixtures\Preload\Process_Wrapper;
+use WP_Rocket\Preload\Homepage;
+
+/**
+ * @covers \WP_Rocket\Preload\Homepage::preload
+ * @group Preload
+ */
+class Test_Preload extends TestCase {
+	protected $identifier         = 'rocket_preload';
+	protected $option_hook_prefix = 'pre_get_rocket_option_';
+	protected $homeWpOption;
+	protected $preloadErrorsTransient;
+	protected $preloadRunningTransient;
+	protected $process;
+
+	public static function setUpBeforeClass() {
+		require_once WP_ROCKET_TESTS_FIXTURES_DIR . '/Preload/Process_Wrapper.php';
+	}
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->homeWpOption            = get_option( 'home' );
+		$this->preloadErrorsTransient  = get_transient( 'rocket_preload_errors' );
+		$this->preloadRunningTransient = get_transient( 'rocket_preload_running' );
+		$this->process                 = new Process_Wrapper();
+
+		update_option( 'home', 'https://example.com/' );
+	}
+
+	public function tearDown() {
+		parent::tearDown();
+
+		if ( $this->process ) {
+			// Added by \WP_Async_Request.
+			remove_action( 'wp_ajax_' . $this->identifier, [ $this->process, 'maybe_handle' ] );
+			remove_action( 'wp_ajax_nopriv_' . $this->identifier, [ $this->process, 'maybe_handle' ] );
+			// Added by \WP_Background_Process.
+			remove_action( $this->identifier . '_cron', [ $this->process, 'handle_cron_healthcheck' ] );
+			remove_filter( 'cron_schedules', [ $this->process, 'schedule_cron_healthcheck' ] );
+		}
+
+		foreach ( [ 'manual_preload', 'cache_mobile', 'do_caching_mobile_files' ] as $option ) {
+			remove_filter( $this->option_hook_prefix . $option, [ $this, 'return_0' ] );
+			remove_filter( $this->option_hook_prefix . $option, [ $this, 'return_1' ] );
+		}
+
+		remove_filter( $this->option_hook_prefix . 'cache_reject_uri', [ $this, 'return_empty_array' ] );
+
+		update_option( 'home', $this->homeWpOption );
+		set_transient( 'rocket_preload_errors', $this->preloadErrorsTransient );
+		set_transient( 'rocket_preload_running', $this->preloadRunningTransient );
+
+		$this->homeWpOption            = null;
+		$this->preloadErrorsTransient  = null;
+		$this->preloadRunningTransient = null;
+		$this->process                 = null;
+	}
+
+	public function testShouldPreloadWhenValidUrls() {
+		$home_urls = [
+			[ 'url' => 'https://example.com/' ],
+			[ 'url' => 'https://example.com/foobar/' ],
+			[ 'url' => 'https://example.com/category/barbaz/', 'mobile' => 1 ],
+		];
+
+		// Fake the requests.
+		Functions\when( 'wp_remote_get' )->alias( function( $url, $args = [] ) {
+			switch ( \trailingslashit( $url ) ) {
+				case 'https://example.com/':
+					$file_name = 'home';
+					break;
+				case 'https://example.com/foobar/':
+					$file_name = 'foobar';
+					break;
+				case 'https://example.com/category/barbaz/':
+					$file_name = 'category-barbaz';
+					break;
+				default:
+					return new \WP_Error( 'wrong-url', 'Wrong URL', [ $url ] );
+			}
```

Comment:
@Screenfeed We have a virtual filesystem now that can be used to build up the cache directory and files. Instead of faking the remote get, instead use the virtual filesystem, save the fixture content into each mobile cache file, and then let the code run as it normally would.

Why? Faking here might mask a problem. By using the virtual filesystem, you are adding cached mobile files and allowing the code to run normally.

Reply:
@hellofromtonya OK for the filesystem, but what about the request itself? As I expected, I get `cURL error 7`.